ddtrace/internal/writer.py

```diff
@@ -116,6 +125,100 @@ def __repr__(self):
         )
 
 
+@attr.s
+class Stitcher(object):
+    _traces = attr.ib(factory=dict, init=False)  # type: List[_Trace]
+
+    @attr.s()
+    class _Trace(object):
+        """Maintains the state of a trace (a list of spans).
+
+        This data structure is required since there are data that apply to an
+        entire trace. Currently this data is passed to the Datadog backend
+        through trace chunk roots (first span of a trace payload to the agent).
+
+        Unfortunately due to this design, trace state has to be maintained within
+        the client and hence this data structure.
+        """
+
+        trace_id = attr.ib(type=int)  # type: int
+        _spans = attr.ib(default=attr.Factory(list))  # type: List[Span]
+        _num_finished = attr.ib(type=int, default=0)  # type: int
+
+        def __len__(self):
+            return len(self._spans)
+
+        def add_span(self, span):
+            # type: (Span) -> None
+            self._spans.append(span)
+
+        def finish_span(self, span):
+            # type: (Span) -> (List[Span], bool)
+            self._num_finished += 1
+            if self._num_finished == len(self._spans) or (
+                partial_flush_enabled and self._num_finished >= partial_flush_min_spans
+            ):
+                finished = [s for s in self._spans if s.finished]
+                self._spans = [s for s in self._spans if not s.finished]
+                self._num_finished -= len(finished)
+                sampled = any(s.sampled for s in finished)
+                return finished, sampled
+            return [], False
+
+    def _get_or_create_trace(self, trace_id):
+        # type: (int) -> Stitcher._Trace
+        if trace_id in self._traces:
+            return self._traces[trace_id]
+        else:
+            trace = self._Trace(trace_id=trace_id)
+            self._traces[trace_id] = trace
+            return trace
+
+    def add_span(self, span):
+        # type: (Span) -> None
+        trace = self._get_or_create_trace(span.trace_id)
+        trace.add_span(span)
+
+    def finish_span(self, span):
+        # type: (Span) -> Tuple[List[Span], bool]
+        trace = self._get_or_create_trace(span.trace_id)
+        spans, sampled = trace.finish_span(span)
+        if spans and len(trace) == 0:
+            del self._traces[span.trace_id]
+        return spans, sampled
+
+
+class BatchWriter(object):
```

Comment:
is there a world where we don't want a `BatchWriter`?

maybe in the future if/when the agent supports streaming, but this isn't something a customer will ever configure, is it?

Reply:
Configuring the writer is actually not in the public API so we don't officially support users configuring a writer. The idea here is to keep the design modular so that we could easily change it out if necessary (eg if span streaming is available and we want to downgrade to this approach).

Reply:
This concept has evolved into being a `SpanProcessor`. As for configuration, it's probably a rare use-case that a user would want to have their own span processor but it should be easy enough to make it configurable. Not sure if we want to do this right away or not though.